Add ETH beneficiary to contributors

Doy-lee opened the pull request:
The necessary oxen-core updates that pairs with https://github.com/oxen-io/eth-sn-contracts/pull/87

Seems to be working. The new contracts emit a `NewServiceNodeV2` which we parse to get the new beneficiary information. When payments are done we ensure we assign a reward to the beneficiary, not, the staker's address. 

In order to make this backwards compatible I've bumped the version of the SNL so on launch everyone will recalculate the SNL and augment all contributors with a beneficiary that is simply the same address as the contributor. Since all the code will use the new structure with beneficiaries we can start paying out to "beneficiaries" even with v1 `NewServiceNode` events.

The only thing that is gated is when it's acceptable to start stuffing V2 `NewServiceNode` data into transactions which affect consensus. For this we will hardfork the network, then, once everyone has upgrade and has the capability to parse the V2 blobs, then, we upgrade the contract so that it starts emitting the new contributor layout.

Earlier I made the wrong assumption that we need to gate the serialisation of the blob by hardfork, that was incorrect and would introduce consensus issues. Universally, when the contract is updated- no node on the network is going to disagree on what version they saw, it'll always be consistently V1 or V2 blobs, never inconsistent between each other so there's no real problem there.

Doy-lee committed:
Add separate beneficiary to ethereum payouts

This change affects consensus by adding a new beneficiary field to the
contributors of an Ethereum contributor in the
`eth::event::NewServiceNode` which is a `tx_extra` field which gets
hashed and hence must be agreed upon by all nodes on the network to be
added to the blockchain.

All existing nodes will have their beneficiary defaulted to the wallet
that staked the contribution hence it is backwards compatible for other
parts of the code to use the new beneficiary field without a hardfork
such as the SNL and SQL DB.

The SNL will recalculate the metadata to ensure that all old nodes will
have their beneficiary set to the wallet's address.

When we intend to hardfork, we will pause the smart contracts, upgrade
them and keep the contracts paused until the network has transitioned
over. This ensures all nodes have upgraded and can parse the new
contract payloads.

Doy-lee committed:
Separate the new fields in v2 to separate L2 structures

Doy-lee committed:
Resolve HF consensus on New SN problem

Instead of trying to retrofit the NewServiceNode emitted log to store
both a V1 and V2 payload, simply introduce a new V2 log that is handled in
a separate step from the V1 payload.

All nodes will observe and agree that a log is V2 when the contract is
upgraded. The contract is _not_ upgraded until all nodes on the network
have hardforked. Thus nodes will always be in agreement of the canonical
representation of the log.

Doy-lee committed:
Update local-devnet tests to test beneficiary and staker unstake payments

Doy-lee committed:
Fix BLS add/sub discarding calculations breaking key aggregator

Doy-lee committed:
Update local devnet to use new ServiceNodeRewards ABI

jagerman commented on src/cryptonote_config.h:
```diff
@@ -252,6 +252,7 @@ enum class hf : uint8_t {
     hf19_reward_batching,
     hf20_eth_transition,  // Temp period: registrations disabled, BLS pubkeys in proofs
     hf21_eth,             // Full transition: registrations from ETH
+    hf22_eth_beneficiary, // Add beneficiary to eth::event::Contributor (and into TX hashes which affects consensus)
```
I don't really want to define a new hardfork for this; our hardforks generally are there to track changes on mainnet consensus, with devnet/testnet/etc. working around it -- but this feature is going to be something launching on the initial mainnet, which means mainnet won't have a meaningful hf21.

Given the plan to put out a mandatory stagenet upgrade, and only afterwards follow up with the contract update that will produce the new values, I don't see a problem (for stagenet) in introducing these retroactively into hf21.  Yes, that's a concensus breakage as soon as the first v2 one hits, but on stagenet, that's okay given the nature of the network.

Doy-lee replied:
Actually yeah, I think we can omit this, even so in mainnet because you'd want to issue a MSNU that bumps the patch not a hardfork and it _should_ be impossible to observe a v2 TX until the contract is upgraded which is done after the HF

jagerman commented on src/l2_tracker/events.h:
```diff
@@ -72,6 +74,63 @@ struct NewServiceNode : L2StateChange {
+    template <class Archive>
+    void serialize_object(Archive& ar) {
+        auto version = tools::enum_top<Version>;
+        field_varint(ar, "version", version, [](auto v) { return v < Version::_count; });
```
Versioning each individual contributor doesn't feel right to me in terms of serialization, because these are never serialized independently of their containing NewServiceNodeV2 which should be where the serialization version gets managed (once).

jagerman replied:
... But I can't think of a simple way to make that work here and allow future Contribution modifications, either.

Doy-lee replied:
Yep you would need a new struct most of the time to handle changes in the structure version would be good to avoid that code explosion.

jagerman committed:
Add support for serialization of std::optionals

For json serialization these become either the value, or the json `null`
literal.

For binary serialization these become a serialized bool (0 or 1)
indicating whether the value is set and then, if (and only if) 1, the
serialized value.  Deserializing loads the bool and then either clears
the optional (if 0) or emplaces and loads it (if 1).

jagerman commented on src/l2_tracker/events.h:
```diff
@@ -72,6 +74,63 @@ struct NewServiceNode : L2StateChange {
+        auto version = tools::enum_top<Version>;
+        field_varint(ar, "version", version, [](auto v) { return v < Version::_count; });
+        field(ar, "address", address);
+        field(ar, "beneficiary", beneficiary);
```
I've created PR #1733 to support serialization of optionals so that we can make this serialization more efficient by making `beneficiary` an std::optional that we can set to nullopt when the address and beneficiary are equal (requiring only 1 byte to serialize in the vast majority of cases where these won't differ).

jagerman replied:
A couple of ways to go with this: making the member variable an optional with a getter for the beneficiary logic:
```C++
std::optional<eth::address> _beneficiary;
const eth::address& beneficiary() const { return _beneficiary ? *_beneficiary : address; }
```
or keeping it as a plain value and dealing with the optional deduplication inside the serialization function:
```C++
std::optional<eth::address> serialized_beneficiary;
if (Archive::is_serializer && beneficiary != address)
    serialized_beneficiary = beneficiary;
field(ar, "beneficiary", serialized_beneficiary);
if (Archive::is_deserializer)
    beneficiary = serialized_beneficiary ? *serialized_beneficiary : address;
```

Doy-lee replied:
Done. I rebased onto #1733 to get the code onto this branch so that we can use your improvements.

jagerman commented on src/l2_tracker/events.h:
```diff
@@ -72,6 +74,63 @@ struct NewServiceNode : L2StateChange {
+    template <class Archive>
+    void serialize_object(Archive& ar) {
+        auto version = tools::enum_top<Version>;
+        field_varint(ar, "version", version, [](auto v) { return v < Version::_count; });
```
This is allowing 0 == `version_invalid` as a valid de-serializable value.

Doy-lee replied:
Fixed

jagerman commented on src/l2_tracker/events.h:
```diff
@@ -72,6 +74,63 @@ struct NewServiceNode : L2StateChange {
     static constexpr std::string_view description = "new SN"sv;
 };
```
Is to plan to get rid of the NewServiceNode and Contributor structs above this?  There is a whole lot of duplication here between the old ones and the new ones.

Thinking ahead of the implications across a hardfork, I don't like the precedent this establishes that every time we change an event we're going to need to duplicate an event struct (and add a txtype, and add separate handling) like this because those structs can never go away, and are duplicating rather a lot.

I'm thinking instead that we could get away with just one NewServiceNode struct here that uses the version to distinguish between contributor types and auto-converts between old contributor values and new ones depending on the version:
```c++
struct Contributor_old {
    eth::address address;
    uint64_t amount;
    
    Contributor_old(const eth::address& addr, uint64_t amt) : address{addr}, amount{amt} {}
    explicit Contributor_old(const Contributor& c) : address{c.address}, amount{c.amount} {}

    template <class Archive>
    void serialize_object(Archive& ar) {
        field(ar, "address", address);
        field_varint(ar, "amount", amount);
    }

    operator Contributor() const {
        return {address, address, amount};
        //return {address, std::nullopt, amount}; // with std::optional suggestion
    }
};

struct NewServiceNode {
    enum class Version { invalid = -1, v0, v1, _count };
    Version version = Version::v1;
    // ...
    void serialize_object(Archive& ar) {
        field_varint(ar, "version", version, [](auto v) { return v > Version::invalid && v < Version:_count; });
        field_varint(ar, "chain_id", chain_id);
        field_varint(ar, "l2_height", l2_height);
        field(ar, "service_node_pubkey", sn_pubkey);
        field(ar, "bls_pubkey", bls_pubkey);
        field(ar, "signature", ed_signature);
        field_varint(ar, "fee", fee);
        if (version == Version::v0) {
            std::vector<Contributor_old> contrib;
            if (Archive::is_serializer)
                contrib = {contributors.begin(), contributors.end()};
            field("contributors", contrib);
            if (Archive::is_deserializer) {
                contributors = {contrib.begin(), contrib.end()};
        } else {
            field("contributors", contributors);
        }
    }

    // I know I've said I don't want to actually do a HF22 for this change, but if we *did* do a HF22
    // with a change like this, this is how we could deal with that conceptually without needing to
    // make a new tx type every time we change something in an event.  (There would need to
    // be some other parts that need to change block acceptance to call this as well):

    // Checks whether the deserialized object is acceptable for the given HF
    bool check_hf_validity(cryptonote::hf hf_version) const {
        if (hf_version < hf::hf21_eth_bls)
            return false;
        if (hf_version < hf::hf22_beneficiaries)
            for (auto& c : contributors)
                if (c.beneficiary != c.address)
                    return false;
        return true;
    }
};
```

Doy-lee replied:
Yeah there's explosion in code with subsequent versions since there's a bunch of places you have to update. Let's avoid that and we can do that with the `version` field as you have in V2 and in the contract's log also emit a version for that event.

With a version we won't need new TXs types for modifying fields, we can just conditionally serialise based on the version on the log. It's either you have the code to serialise or or you won't in which case you need to upgrade your node (e.g. a MSNU that bumps the patch not hardfork) So to patch this up, I'll add a version to the smart contract log and the struct here.

With `check_hf_validity` we want to avoid using a HF value to decide how to serialise/validate TX extra's because every node synchronises the L2 TXs independently from each other. If a node syncs a L2 transaction into the mempool on an Oxen  HF boundary, some could serialise it as V1 if they are slow to HF and some could serialise it as V2 if they observed the HF and this would cause consensus issues between the two partitions.

All the V1s will be removed before mainnet and to relaunch stagenet without the legacy data structures. 

So I think this is sorted with a version field. To retrofit it into `NewServiceNodeV1` is a bit tedious to handle (how to figure out if the version is present and at the front of the payload or not?) and something that we're going to drop so if it's ok, I'd keep the V1s around and drop them by launch.

jagerman commented on src/cryptonote_basic/txtypes.h:
```diff
@@ -28,12 +28,13 @@ enum class txtype : uint16_t {
     ethereum_service_node_exit_request,
     ethereum_service_node_exit,
     ethereum_staking_requirement_updated,
+    ethereum_new_service_node_v2,
```
I'd like to try to avoid needing to use new tx types for different versions of the same conceptual tx type.  See my other, longer comment about how we might do that.

Doy-lee replied:
Similar as above, it's temporary and to be dropped by launch. If we were to get rid of it now I guess is what's the way to retroactively figure out if the NewSN payload has a version or not (maybe a magic value in the header?)

Doy-lee committed:
Disallow invalid version in Contributor serialisation

Doy-lee committed:
Remove HF22, retroactively support NewSNv2

Doy-lee committed:
Use optional to serialise beneficiary

Doy-lee committed:
Add version to top of NewServiceNodeV2 struct and log

Doy-lee committed:
Ensure version is encoded in ABI and handled correctly

Doy-lee committed:
Handle NewSNv2 in cn_deserialize

Doy-lee committed:
Format code to appease lint checker

Doy-lee commented:
Merged in https://github.com/oxen-io/oxen-core/pull/1746
